### ISSUE TITLE
MISRA 2023 standards missing descriptions

### ISSUE THREAD

**Issue by nstander** [stinb/plugins#23]

**Comment by groke** [comment]
Added with https://github.com/stinb/sti/commit/4220b5e9c08295afb2169e75f67b720921a3fc3d

**Comment by katescitools** [comment]
Started testing this today. Kindly see below for some concerns: @nstander 

**Issue 1:** I can still see some checks missing description:
MISRA23_10.4

![image](https://github.com/user-attachments/assets/dfc9d3ca-f455-462e-9bce-c51942c337b5)

MISRA23_10.6

![image](https://github.com/user-attachments/assets/49bff807-6ca7-43a6-9a6b-4fd5b33adb0e)

**Issue 2:** Is Amplification intended to be here:

MISRA23_11.1

![image](https://github.com/user-attachments/assets/0cd6801e-4905-477a-afb7-e99b9ef00505)

**Issue 3:** (Clarification needed) I’m assuming that this issue is only particular to Description found in the check’s info, and that checks that contain data like this are expected. Kindly confirm:

![image](https://github.com/user-attachments/assets/dbf70e3a-b5e9-4a25-9670-ffab60422597)

**Issue 4:** (Clarification needed) Is a description like this intended for some checks?

![image](https://github.com/user-attachments/assets/180b2233-68ec-4c86-890c-504c5e5f9921)

Also this: 

![image](https://github.com/user-attachments/assets/09a3a72b-62a3-425d-b1cd-136e0d70d92b)

**Comment by jordancolbeth** [comment]
@katescitools 12.1 and 20.3 are no-op checks, meaning they are either not implemented at all or they are handled by the clang parser. Those checks are expected to have the line mentioning either that the check cannot be implemented, or that it's automatically run. That said, 12.1 looks to be missing a large chunk of the description so I'll get that added. 20.3 looks ok to me. 

@jasonbquinn @RobbyCBennett would you please go through and fix the descriptions for whichever checks have your initials next to them below?

- [ ] MISRA23_10.4 - JQ
- [ ] MISRA23_10.6 - JQ
- [ ] MISRA23_11.1 - RB
- [ ] MISRA23_15.2 - JC
- [ ] MISRA23_12.1 - JC

**Comment by kylejacksonb** [comment]
MISRA23_11.1 had a formatting error in its description. I updated it so now it renders the same format as MISRA23_10.4 and MISRA23_10.6.

For MISRA23_10.4 and MISRA23_10.6, their descriptions start with 'Amplification' in the official documentation.

For MISRA23_12.1 and MISRA23_20.3 ( And pretty much every other NO-OP check defined in a JSON file ) they use HTML markup that follows this pattern:
```
"<p><b>",
"This check cannot be automated",
"</b></p>",
```
Unless I miss my guess, CodeCheck is looking for this pattern to decide what gets its own header. We can swap the closing bold and paragraph tags so that it no longer has a dropdown. ( This was the issue with MISRA23_11.1 ). 

@groke @HeidiJoe This is mostly a design question. How do you want this information to look when the description starts with 'Amplification' and are you happy with how NO-OP checks are rendering?

**Comment by nstander** [comment]
> Unless I miss my guess, CodeCheck is looking for this pattern to decide what gets its own header.

Yes. The `<p><b>` and `</b></p>` determines what gets formatted as headers.

**Comment by kylejacksonb** [comment]
@RobbyCBennett @jasonbquinn Let's hold off until we get clarification from Kevin or Heidi.

**Comment by groke** [comment]
How does it look if we bold the word Amplification but don't wrap it in <p> so it doesn't end up as a seperate header?

**Comment by kylejacksonb** [comment]
![image](https://github.com/user-attachments/assets/5ec14e9a-f4ca-4d33-8f9f-7147b7b51a5a)

**Comment by groke** [comment]
We don't want any blank sections, so just treat Amplification like a regular header and don't do a description if it is blank

**Comment by jordancolbeth** [comment]
@groke The issue is that the amplification, rationale, and exceptions are subheaders of the description, so the description should never be empty. If I'm understanding you correctly are you saying that unless it's a no-op the description header should be removed?

**Comment by RobbyCBennett** [comment]
The description header is implicitly created because the first element is a plain old tag. To fix mine, I removed the first tag which was `<p></p>`

**Comment by nstander** [comment]
> The description header is implicitly created because the first element is a plain old paragraph tag (or some other non-bold thing)

Yes, the Ui creates a "Description" header if there is any non-whitespace characters that occur before the first `<p><b>` heading flag.

**Comment by jordancolbeth** [comment]
So, it's not just 12.1 that's missing the full description, the large majority of no-ops are. These take time to implement, do we want to make that a separate issue/project for later?

**Comment by kylejacksonb** [comment]
Yeah, let's keep this issue focused on fixing formatting.

**Comment by RobbyCBennett** [comment]
> I think that'll do it. Any thoughts or premonitions of breaking everything?

Change 4 will leave open tags in some cases like in CPP_D009. Also make sure to look at the rendered descriptions after changing them.

**Comment by kylejacksonb** [comment]
CPP_D009 has this:
```
<p><b>See also</b><br>
ISO/IEC 14882:2003 [1] §7.3.3</p>
```
Without the extra opening tag at the end of the replace, then we have an extra closing tag.

**Comment by RobbyCBennett** [comment]
My bad. I didn't see that in the mess of symbols. I guess an empty tag is unlikely and also it's probably going to be rendered fine anyway.

**Comment by kylejacksonb** [comment]
@jasonbquinn You have the most experience with writing scripts to do mass updates on CodeCheck. Would you please put together a script for task 1 and 4?

**Comment by katescitools** [comment]
@nstander , just to confirm as there's no description provided for this issue, do you mean we're only supposed to check the descriptions found in CodeCheck Documentation (for whole standard) like this?

![image](https://github.com/user-attachments/assets/0d81e6d7-94cc-4aef-b3e6-16d03e196287)

If that's the case, MISRA C and C++ 2023 are confirmed to have their descriptions. However, I noticed a minor issue: the statement 'This check is automatically run' is positioned too close to the groupbox/gray area. (This issue is also replicable for similar bold statements in other checks.) I'm not sure though if it's because of the recent fix @jasonbquinn since they are still considered individual checks, just found in CodeCheck Documentation.

![image](https://github.com/user-attachments/assets/795a2428-7201-420a-9765-87014292dc5c)

It looks like the space/padding is broken because it appeared fine before the fix.

![image](https://github.com/user-attachments/assets/9f2f1849-f803-4d3f-8153-8406b4dbe890)

Before:

![image](https://github.com/user-attachments/assets/a57daa46-b33b-474c-b552-8f4fa9963fd4)

After:

![image](https://github.com/user-attachments/assets/e73c5e6d-77aa-4819-af4f-38fb117a8a5a)

For previous items I raised, I've confirmed that:

1. Descriptions are removed when its value is blank **(PASSED)**
2. "This check cannot be automated" is now changed to just bold, instead of being a separate header *(PASSED)**
3.  "This check is automatically run" is now changed to just bold, instead of being a separate header *(PASSED)**
4.  See also is still found in some check's description: *(FAILED)**

MISRA23_3.2
![image](https://github.com/user-attachments/assets/89c75545-2378-4839-aa07-dd725183c26a)


5. There are still some checks where Amplification is not treated as header: **(FAILED)**

MISRA23_8.5
![image](https://github.com/user-attachments/assets/126d98b3-b773-4c40-bbd5-3686ab3b1ec6)

6. For format consistency, I believe "Example" should also be a separate header. **(FAILED)**
Here's a sample where it's a separate header:

![image](https://github.com/user-attachments/assets/7c00f00b-856f-431c-a12d-b6f4a8fc8185)

"Example" in some checks:

MISRA23_9.2
![image](https://github.com/user-attachments/assets/002af87a-4393-4fba-aea3-705cec7076ce)

Check with similar issue:
MISRA23_9.3
MISRA23_8.19.1
MISRA23_10.1.1
MISRA23_19.2.2

7. This header is a bit weird: (looks like it's not supposed to be there or should be replaced as "Exception") **(FAILED)**

MISRA23_19.0.2
![image](https://github.com/user-attachments/assets/35be9640-eabf-42be-b04d-129c7c510ef4)


**NOTE:** I only checked MISRA 2023 (C and C++), let me know if I should check all standards @kylejacksonb . Also, I didn't create a separate issue for no-op check as I'm not certain of the details, but let me know if I should create one general issue for it.

**Comment by nstander** [comment]
> do you mean we're only supposed to check the descriptions found in CodeCheck Documentation (for whole standard) like this

Yes, that is the only thing needed for the initial issue.

> It looks like the space/padding is broken because it appeared fine before the fix.

If we want the spacing restored, it should be changed at the report level rather than individual checks.

**Comment by kylejacksonb** [comment]
@katescitools these issues have been fixed. And yes, let's start working through the other standards as well.

As for the no-op checks, what are you seeing?

**Comment by katescitools** [comment]
Thanks for the fix, @jasonbquinn ! I've confirmed that all of them are now showing fine in MISRA 2023. 

@kylejacksonb , for no-op checks, I'm referring to this convo:

![Image](https://github.com/user-attachments/assets/4bbbf744-81bc-4cff-a4e0-4ed4f05eab4a)

I will start working on other standards once I've completed the BVT for 1216, probably next week! I'll update here accordingly.

**Comment by kylejacksonb** [comment]
@katescitools Since the original issue is fixed and I think all the formatting issues with 2023 have been fixed, I think we're good to close this issue. I already created ( https://github.com/stinb/plugins/issues/32 ) for the no-op checks. 

Let's make another issue for the other standards if more formatting issues come up. I know Quinn made a bunch of changes to other standards already, so hopefully any additional issues will be few and far between.

**Comment by katescitools** [comment]
Thanks for letting me know, @kylejacksonb . I've created a separate issue (https://github.com/stinb/plugins/issues/36) for other standards with formatting issues. I'll close this issue as well :)
